common/common.go

```diff
@@ -33,3 +39,33 @@ func (r Role) String() string {
 	}
 	return "standby"
 }
+
+// Write file to temp and atomically move when everything else succeeds.
+// This function is taken from
+//   https://github.com/youtube/vitess/blob/master/go/ioutil2/ioutil.go
+// Copyright 2012, Google Inc. BSD-license
+func WriteFileAtomic(filename string, data []byte, perm os.FileMode) error {
+	dir, name := path.Split(filename)
+	f, err := ioutil.TempFile(dir, name)
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(data)
+	if err == nil {
+		err = f.Sync()
+	}
+	if closeErr := f.Close(); err == nil {
+		err = closeErr
+	}
+	if permErr := os.Chmod(f.Name(), perm); err == nil {
+		err = permErr
+	}
+	if err == nil {
+		err = os.Rename(f.Name(), filename)
+	}
+	// Any err should result in full cleanup.
+	if err != nil {
+		os.Remove(f.Name())
```

Comment:
`defer` adds the function call to the list of things to do upon exiting. Here, `os.Remove` is invoked at the very end, so using `defer` does not bring any benefits.

I kept the original version from the upstream project hence the somewhat puzzling code flow (i.e. unusual error handling strategy) but the code seems just fine otherwise.